implementations/rust/kex/src/xx.rs

```diff
@@ -106,31 +111,43 @@ impl KeyExchange for SymmetricState {
         &mut self,
         secret_handle: SecretKeyContext,
         public_key: PublicKey,
-    ) -> Result<Vec<u8>, VaultFailError> {
+    ) -> Result<Vec<SecretKeyContext>, VaultFailError> {
         let mut vault = self.vault.lock().unwrap();
+
+        let attributes_ck = SecretKeyAttributes {
+            xtype: SecretKeyType::Buffer(32),
+            purpose: SecretPurposeType::KeyAgreement,
+            persistence: SecretPersistenceType::Ephemeral,
+        };
+
+        let attributes_k = SecretKeyAttributes {
+            xtype: SecretKeyType::Aes256,
+            purpose: SecretPurposeType::KeyAgreement,
+            persistence: SecretPersistenceType::Ephemeral,
+        };
+
         vault.ec_diffie_hellman_hkdf_sha256(
             secret_handle,
             public_key,
-            self.state.ck.as_ref(),
-            SHA256_SIZE + AES256_KEYSIZE,
+            self.state.ck.unwrap(),
+            vec![attributes_ck, attributes_k],
         )
     }
 
     /// mix key step in Noise protocol
-    fn mix_key<B: AsRef<[u8]>>(&mut self, hash: B) -> Result<(), VaultFailError> {
-        let hash = hash.as_ref();
-        self.state.ck.copy_from_slice(&hash[..SHA256_SIZE]);
-        let attributes = SecretKeyAttributes {
-            xtype: SecretKeyType::Aes256,
-            persistence: SecretPersistenceType::Ephemeral,
-            purpose: SecretPurposeType::KeyAgreement,
-        };
-        let key = SecretKey::Aes256(*array_ref![hash, SHA256_SIZE, AES256_KEYSIZE]);
+    fn mix_key(&mut self, hkdf_output: Vec<SecretKeyContext>) -> Result<(), VaultFailError> {
         let mut vault = self.vault.lock().unwrap();
+
+        if self.state.ck.is_some() {
+            vault.secret_destroy(self.state.ck.unwrap())?;
+        }
+        self.state.ck = Some(hkdf_output[0]);
+
         if self.key.is_some() {
             vault.secret_destroy(*self.key.as_ref().unwrap())?;
         }
-        self.key = Some(vault.secret_import(&key, attributes)?);
+        // FIXME
+        self.key = Some(hkdf_output[1]);
```

Comment:
again here, prefer `.get(1)` here and handle the Option.. could just assign `self.key = hdkf_output.get(1);` directly, but unsure if `self.key` is checked later or if it would be better to catch an unexpected state here.